Review1

OrlovAlexey opened the pull request:
Ну что-то я добавил классы-шифры, как Вы сказали, но все перестало работать(программа ошибок не выдает, но получается не шифр, а пустая строка). Вероятно, проблема в том, что абстрактный шифр не хочет к обычным кастоваться, но это не точно.

al3x3y0rl0v committed:
Review1 task

al3x3y0rl0v committed:
Update README.md

al3x3y0rl0v committed:
Update README.md

al3x3y0rl0v committed:
Little changes

al3x3y0rl0v committed:
Added cipher classes

pkositsyn commented on review1_encryption.py:
```diff
@@ -0,0 +1,248 @@
+
+
+alphabet_rus = "абвгдеёжзийклмнопрстуфхцчшщъыьэюя"
+alphabet_eng = [letter for letter in string.ascii_lowercase]
```
Чем тебя строка не устроила?)

OrlovAlexey replied:
ну так понятнее, что у нас именно буквы из алфавита, а не какая-то строка непонятная. Исправил

pkositsyn commented on review1_encryption.py:
```diff
@@ -0,0 +1,248 @@
+    def __init__(self, text):
+        self.txt = text
+
+def caesar(input_str, bias = 3, language = "eng"):
```
Довольно странное значение 3, я бы вообще без дефолта сделал

OrlovAlexey replied:
так обычно в цезаре сдвиг 3 равен, или я что-то путаю

pkositsyn commented on review1_encryption.py:
```diff
@@ -0,0 +1,248 @@
+        key_str = str(input("Enter your keystring, please: "))
+        return vernam_decryption(input_str, keystr)
+
+class Frequency_analysis(AbstractCipher):
```
Ну это не шифр, не стоило его подгонять под интерфейс, лучше уж отдельную ветку кода для этого сделать

OrlovAlexey replied:
вроде исправил

pkositsyn commented on review1_encryption.py:
```diff
@@ -0,0 +1,248 @@
+        return None
+
+    def decrypt(self, input_str):
+        return frequency_analysis(input_str, self.language)
```
Давай вот эти все функции и внесем в нужные классы, незачем их глобально держать

OrlovAlexey replied:
внес

pkositsyn commented on review1_encryption.py:
```diff
@@ -0,0 +1,248 @@
+
+    with open(args.input_file, 'r') as input_file:
+        input_str = input_file.read()
+    with open(args.mode + ".py", 'w') as output_file:
```
Давай его открывать в более узком скоупе, только там где надо что-то записать

pkositsyn replied:
Для этого лучше разделить части с записью в файл и получением зашифрованных данных

pkositsyn replied:
А почему от .py?)

OrlovAlexey replied:
От .py потому что в args.mode лежит либо "encryption" либо "decryption", соответственно названия файлов куда запишется будут такими же

OrlovAlexey replied:
разделил запись и остальное

OrlovAlexey replied:
А, до меня дошло, конечно не .py, а .txt должно быть))

pkositsyn commented on review1_encryption.py:
```diff
@@ -0,0 +1,248 @@
+        key_str is randomly generated key THE SAME SIZE AS input_str
+    """
+    print(len(input_str))
+    print(len(key_str))
```
Вроде уже писал что принты - не самая лучшая идея

OrlovAlexey replied:
так, я чето пропустил про принты, а как их тогда заменить, чтобы с пользователем можно было в консоле общаться

OrlovAlexey replied:
в поток выводить какой?

OrlovAlexey replied:
а, это принты для дебага были вообще, убрал

pkositsyn commented on review1_encryption.py:
```diff
@@ -0,0 +1,248 @@
+    if language == "eng":
+        alphabet = alphabet_eng
+    elif language == "rus":
+        alphabet = alphabet_rus
```
Вот это бы в класс унести

OrlovAlexey replied:
ну унес

pkositsyn commented on review1_encryption.py:
```diff
@@ -0,0 +1,248 @@
+        super().__init__()
+        self.language = language
+
+    def encrypt(self, input_str):
```
type annotations бы

OrlovAlexey replied:
добавил

pkositsyn commented on review1_encryption.py:
```diff
@@ -0,0 +1,248 @@
+        language = {"eng", "rus"}
+    """
+    
+    if language == "eng":
+        alphabet = alphabet_eng
+        most_commoly_used_letter_index = alphabet.index('e')
+    elif language == "rus":
+        alphabet = alphabet_rus
+        most_commoly_used_letter_index = alphabet.index('о')
```
Для допбалла тут надо обучить анализ на тексте (любом), приближая реальное распределение букв, а потом сравнивать их по какой-то метрике. Текущий вариант, ну на троечку

OrlovAlexey replied:
а обучать надо при каждом запуске получается надо? просто это долго получается

OrlovAlexey replied:
а, ну можно только при первом запуске обучать, сохранять распределения букв в отдельный файл, а потом считывать оттуда, наверно так

OrlovAlexey replied:
ну сделал

pkositsyn reviewed (changes requested):
Если сделаешь O(1) доп памяти на шифрование (сейчас загружаешь разом весь текст), то поставлю +1 балл, а так просто с исправлениями +0.6

pkositsyn commented:
PS ничего не работает, потому что названия модов не совпадают с ифами, лучше такие вещи в константы выносить

al3x3y0rl0v committed:
1st correction

al3x3y0rl0v committed:
2nd correction

al3x3y0rl0v committed:
3rd correction

al3x3y0rl0v committed:
4th correction

al3x3y0rl0v committed:
5th correction

al3x3y0rl0v committed:
7th correction

al3x3y0rl0v committed:
8th correction

al3x3y0rl0v committed:
Added some constants and made it work

OrlovAlexey commented:
> PS ничего не работает, потому что названия модов не совпадают с ифами, лучше такие вещи в константы выносить

да, исправил

al3x3y0rl0v committed:
Added normal frequency analysis

Also added files in which there are examples of real literature text in eng/rus language

al3x3y0rl0v committed:
Added O(1) extra memory

al3x3y0rl0v committed:
Some minor changes

OrlovAlexey commented:
> Если сделаешь O(1) доп памяти на шифрование (сейчас загружаешь разом весь текст), то поставлю +1 балл, а так просто с исправлениями +0.6

сделал сделал

pkositsyn commented on review1_encryption.py:
```diff
@@ -1,248 +1,312 @@
+alphabet_eng = string.ascii_lowercase
+eng_tag = "eng"
+rus_tag = "rus"
+mode1 = "encryption"
```
Константы капсом обычно в питоне, но ладно

pkositsyn commented on review1_encryption.py:
```diff
@@ -1,248 +1,312 @@
+        output_str = ''.join(new_str)
+        return output_str
+
+def encrypt_block(block: str, mode: str, cipher: AbstractCipher) -> str:
```
Теоретически можно было бы и в абстрактный класс как метод внести, но и так нормально смотрится

pkositsyn reviewed (approved):
Max block size особо не заиспользовал в итоге. Ну и читать вероятно стоило не по строкам а по Х символам. Но исправлено классно, +1 твои
